python_modules/dagster/dagster/_core/definitions/metadata/metadata_set.py

```diff
@@ -82,10 +111,35 @@ class MyMetadataSet(NamedspacedMetadataSet):
             if len(splits) == 2:
                 namespace, key = splits
                 if namespace == cls.namespace() and key in model_fields(cls):
-                    kwargs[key] = value.value if isinstance(value, MetadataValue) else value
+                    kwargs[key] = cls._extract_value(field_name=key, value=value)
 
         return cls(**kwargs)
 
+    @classmethod
+    def _extract_value(cls, field_name: str, value: Any) -> Any:
+        """Based on type annotation, potentially coerce the metadata value to its inner value.
+
+        E.g. if the annotation is Optional[float] and the value is FloatMetadataValue, construct
+        the MetadataSet using the inner float.
+        """
+        if isinstance(value, MetadataValue):
+            annotation = model_fields(cls)[field_name].annotation
+            annotation_acceptable_types = set(flatten_annotation_unions(annotation))
+            if (
+                type(value) not in annotation_acceptable_types
+                and type(value.value) in annotation_acceptable_types
+            ):
+                check.invariant(type(value.value) in DIRECTLY_WRAPPED_METADATA_TYPES)
+                return value.value
+
+        return value
+
+    @classmethod
+    @lru_cache(maxsize=128)
+    def _get_accepted_types_for_field(cls, field_name: str) -> AbstractSet[Type]:
```

Comment:
A bit confused why we are annotating this with `lru_cache` since it is just called once per field in `__init__`. I would suggest removing it.

Reply:
It only needs to be computed once per field per class, but it gets recomputed once per field per instance (+ once per field per `extract` invocation), which feels wasteful. If there was an easy way to do it at class definition time, I would do that instead, but, if there is one, it probably involves fancy metaclass stuff that's not worth messing with. The cache will never get large because it's O(# metadata set classes * # fields).  Maxsize should probably actually be infinite to make this more predictable.

Don't feel strongly though - can remove it if you do.

Reply:
Oh that makes sense. 👍🏻

Reply:
👍 added a comment